portal/src/util/validation.ts

```diff
@@ -0,0 +1,59 @@
+export interface ValidationRule<Data, ErrorState> {
+  inputKey: keyof Data;
+  errorKey: keyof ErrorState;
+  condition: (input: any) => boolean;
+  errorMessageId: string;
+}
+
+export type ErrorResult<ErrorState> = Partial<
+  { [key in keyof ErrorState]: string }
+>;
+
+export function validateInput<Data, ErrorState>(
+  data: Data,
+  validationRules: ValidationRule<Data, ErrorState>[]
+): {
+  errorResult: ErrorResult<ErrorState>;
+  isValid: boolean;
+} {
+  let isValid = true;
+  const errorResult: ErrorResult<ErrorState> = {};
+  validationRules.forEach(
+    (validationRule: ValidationRule<Data, ErrorState>) => {
+      if (!validationRule.condition(data[validationRule.inputKey])) {
+        errorResult[validationRule.errorKey] = validationRule.errorMessageId;
+      }
+      isValid = false;
+    }
+  );
+  return { errorResult, isValid };
+}
+
+export function isValidEmail(email: string): boolean {
+  const emailRegExp = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/;
+  return emailRegExp.test(email) || email === "";
+}
+
+export function isValidEmailDomain(domain: string): boolean {
+  const emailDomainRegExp = /^\w+([.-]?\w+)*(\.\w{2,3})+$/;
+  return emailDomainRegExp.test(domain) || domain === "";
+}
+
+export function isErrorResult<ErrorState>(
+  result: any
+): result is ErrorResult<ErrorState> {
+  if (!(result instanceof Object)) {
+    return false;
+  }
+  return Object.keys(result).length > 0;
+}
+
+export function renderErrorMessage(
```

Comment:
We are going to rely on the backend JSON schema to validation. So we do not need this module at the moment.